### PR TITLE
Tab-complete docker:// URLs where possible.

### LINKED ISSUE #403: Have bash completion for exec of docker://-style URLs
@vsoch asked about bash completion; I thought this would be a nifty little improvement that would demonstrate how to interact with the bash completion.

Basically, I think if we do something along these lines:
```
singularity exec docker://centos:<tab><tab>
```
then bash should lookup a listing of container labels under the `centos` account.  Basically, we'd need to write a script to do the query - and have it invoked by the bash completion script.

### PR DESCRIPTION
**Pull request by bbockelm**
With this, you currently get the following suggested completion:

```
$ singularity exec docker://centos:<TAB><TAB>
5               6.8             7.3.1611        centos6.7       centos7.2.1511
5.11            7               centos5         centos6.8       centos7.3.1611
6               7.0.1406        centos5.11      centos7         latest
6.6             7.1.1503        centos6         centos7.0.1406
6.7             7.2.1511        centos6.6       centos7.1.1503
```

Fixes #403

@singularityware-admin

### CODE REVIEW

File: etc/bash_completion.d/singularity.in [hunk 1]
```diff
@@ -79,8 +85,16 @@ _singularity() {
                 COMPREPLY=( $(compgen -f -- ${cur}) )
             elif [[ ${cur} == -* ]] ; then
                 COMPREPLY=( $(compgen -W "--help --bind --scratch --home --pwd --contain --containall --ipc --pid --user --workdir --writable" -- ${cur}) )
+            elif [[ ${cur} == docker://*:* ]] ; then
+                # By default, bash considers ':' as a word separator.  We must thus strip the prefix docker: from our responses.
+                repo_name=$(echo ${cur} | sed -e 's|docker://\([-a-zA-Z0-9._]*\):.*|\1|')
+                response_var=$(echo ${cur} | sed -e 's|docker://[-a-zA-Z0-9._]*:||')
+                all_tags=$(PYTHONPATH="${exec_prefix}/libexec/singularity/python" python -c 'for i in __import__("docker.api").api.get_tags(namespace="library", repo_name=__import__("sys").argv[1]): print "%s" % i,' ${repo_name} 2>/dev/null)
```
Comment (bbockelm) [inline etc/bash_completion.d/singularity.in]: If this piece of voodoo doesn't make you frustrated with bash programming, I don't know what will!

Comment (vsoch) [reply etc/bash_completion.d/singularity.in]: This is amazing. Testing now!

**Comment by vsoch** [comment]
This is great! One tiny tweak should make this python 2/3 compatible, try like this:

      all_tags=$(PYTHONPATH="${exec_prefix}/libexec/singularity/python" python -c 'for i in __import__("docker.api").api.get_tags(namespace="library", repo_name=__import__("sys").argv[1]): print("%s" % i),' ${repo_name} 2>/dev/null)

**Comment by vsoch** [comment]
the one remaining potential thing to add is compatability with namespaces that aren't library, for example tensorflow (`tensorflow/tensorflow:latest`).  To make this work, we would need one more level of parsing that can also detect a namespace, based on having a "/" first instead of a ":"

User inputs:

       docker://tensorflow/tensorflow:latest

Default namespace is library. But then the we detect a "/" before a ":" after docker, and parse "tensorflow" as the namespace, and the function call becomes:

      all_tags=$(PYTHONPATH="${exec_prefix}/libexec/singularity/python" python -c 'for i in __import__("docker.api").api.get_tags(namespace=__import__("sys").argv[1], repo_name=__import__("sys").argv[2]): print(i),' ${namespace} ${repo_name})

or something like that.

**Comment by vsoch** [comment]
hey @bbockelm ! Here is a start that will get us closer to having tab completion for docker library (standard namespace) and custom registries! What I did is add a new python module called `shell.py` under the docker folder that can take in the entire clump of stuff before the ":", and then parse from that the namespace, and repo_name.

https://github.com/vsoch/singularity/tree/bbockelm-docker_tab_complete

It still works for standard images (like centos) and running the functions alone, the python calls should technically work for tensorflow/tensorflow:, but I think it's hiccuping on something with the bash sniffing part, because the output never gets printed (I think it might be the second / gets read in again, or something like that). If you have some time over break, could you take a look? Once we figure out this last thing, I think the bash completion should work for all different kinds of registries!

**Comment by gmkurtzer** [comment]
Heya guys, what's the status of this?

**Comment by vsoch** [comment]
I had done the python bits to retrieve the tags - I have no idea how the underlying bash works so I stopped and handed it back to @bbockelm. Also - I think the current development completion might be broken... :sob:

**Comment by gmkurtzer** [comment]
How do we receive tags? I remember we used to be able to do that, and it was really cool!

Oh? tab completion is broken now? Uh-oh... What did you find, I will try and replicate.

**Comment by vsoch** [comment]
it's pretty easy to get tags now, even you could do this @gmkurtzer !

```bash
cd singularity/libexec/python
python
```
now we are in python...

```python
from docker.api import *
VERBOSE2 SINGULARITY_COMMAND_ASIS found as False
VERBOSE2 SINGULARITY_ROOTFS not defined (None)
VERBOSE2 SINGULARITY_METADATA_FOLDER found as None/.singularity.d
VERBOSE2 SINGULARITY_DISABLE_CACHE found as False
VERBOSE2 SINGULARITY_CACHEDIR found as /home/vanessa/.singularity
VERBOSE2 SINGULARITY_ENVBASE found as None/.singularity.d/env
VERBOSE2 SINGULARITY_LABELFILE found as None/.singularity.d/labels.json
VERBOSE2 SINGULARITY_INCLUDECMD found as False
VERBOSE2 SINGULARITY_PULLFOLDER found as /home/vanessa/Documents/Dropbox/Code/singularity/singularity/libexec/python
VERBOSE2 SHUB_NAMEBYHASH not defined (None)
VERBOSE2 SHUB_NAMEBYCOMMIT not defined (None)
VERBOSE2 SHUB_CONTAINERNAME not defined (None)
VERBOSE2 SINGULARITY_CONTENTS found as None/.singularity.d/.layers

In [2]: client = DockerApiConnection(image="ubuntu")
DEBUG Headers found: Accept,Content-Type
VERBOSE Registry: index.docker.io
VERBOSE Namespace: library
VERBOSE Repo Name: ubuntu
VERBOSE Repo Tag: latest
VERBOSE Version: None
VERBOSE Obtaining tags: https://index.docker.io/v2/library/ubuntu/tags/list
DEBUG GET https://index.docker.io/v2/library/ubuntu/tags/list
DEBUG Http Error with code 401
DEBUG GET https://auth.docker.io/token?service=registry.docker.io&expires_in=9000&scope=repository:library/ubuntu:pull
DEBUG Headers found: Accept,Content-Type,Authorization

In [3]: tags = client.get_tags()
VERBOSE Obtaining tags: https://index.docker.io/v2/library/ubuntu/tags/list
DEBUG GET https://index.docker.io/v2/library/ubuntu/tags/list

tags
['10.04',
 '12.04.5',
 '12.04',
 '12.10',
 '13.04',
 '13.10',
 '14.04.1',
 '14.04.2',
 '14.04.3',
 '14.04.4',
 '14.04.5',
 '14.04',
 '14.10',
 '15.04',
 '15.10',
 '16.04',
 '16.10',
 '17.04',
 'devel',
 'latest',
 'lucid',
 'precise-20150212',
 'precise-20150228.11',
 'precise-20150320',
 'precise-20150427',
 'precise-20150528',
 'precise-20150612',
 'precise-20150626',
 'precise-20150729',
 'precise-20150813',
 'precise-20150924',
 'precise-20151020',
 'precise-20151028',
 'precise-20151208',
 'precise-20160108',
 'precise-20160217',
 'precise-20160225',
 'precise-20160303',
 'precise-20160311',
 'precise-20160318',
 'precise-20160330',
 'precise-20160425',
 'precise-20160503',
 'precise-20160526',
 'precise-20160624',
 'precise-20160707',
 'precise-20160819',
 'precise-20160923.1',
 'precise-20161102',
 'precise-20161123',
 'precise-20161209',
 'precise-20170214',
 'precise-20170331',
 'precise',
 'quantal',
 'raring',
 'rolling',
 'saucy',
 'trusty-20150218.1',
 'trusty-20150228.11',
 'trusty-20150320',
 'trusty-20150427',
 'trusty-20150528',
 'trusty-20150612',
 'trusty-20150630',
 'trusty-20150730',
 'trusty-20150806',
 'trusty-20150814',
 'trusty-20151001',
 'trusty-20151009',
 'trusty-20151021',
 'trusty-20151028',
 'trusty-20151208',
 'trusty-20151218',
 'trusty-20160119',
 'trusty-20160217',
 'trusty-20160226',
 'trusty-20160302',
 'trusty-20160315',
 'trusty-20160317',
 'trusty-20160323',
 'trusty-20160405',
 'trusty-20160412',
 'trusty-20160424',
 'trusty-20160503.1',
 'trusty-20160526',
 'trusty-20160624',
 'trusty-20160711',
 'trusty-20160802',
 'trusty-20160819',
 'trusty-20160914',
 'trusty-20160923.1',
 'trusty-20161006',
 'trusty-20161101',
 'trusty-20161123',
 'trusty-20161214',
 'trusty-20170119',
 'trusty-20170214',
 'trusty-20170330',
 'trusty',
 'utopic-20150211',
 'utopic-20150228.11',
 'utopic-20150319',
 'utopic-20150418',
 'utopic-20150427',
 'utopic-20150528',
 'utopic-20150612',
 'utopic-20150625',
 'utopic',
 'vivid-20150218',
 'vivid-20150309',
 'vivid-20150319.1',
 'vivid-20150421',
 'vivid-20150427',
 'vivid-20150528',
 'vivid-20150611',
 'vivid-20150802',
 'vivid-20150813',
 'vivid-20150930',
 'vivid-20151021',
 'vivid-20151106',
 'vivid-20151111',
 'vivid-20151208',
 'vivid-20160122',
 'vivid',
 'wily-20150528.1',
 'wily-20150611',
 'wily-20150708',
 'wily-20150731',
 'wily-20150807',
 'wily-20150818',
 'wily-20150829',
 'wily-20151006',
 'wily-20151009',
 'wily-20151019',
 'wily-20151208',
 'wily-20160121',
 'wily-20160217',
 'wily-20160302',
 'wily-20160316',
 'wily-20160329',
 'wily-20160424',
 'wily-20160503',
 'wily-20160526',
 'wily-20160602',
 'wily-20160706',
 'wily',
 'xenial-20151218.1',
 'xenial-20160119.1',
 'xenial-20160125',
 'xenial-20160217.2',
 'xenial-20160226',
 'xenial-20160303.1',
 'xenial-20160314.4',
 'xenial-20160317',
 'xenial-20160331.1',
 'xenial-20160422',
 'xenial-20160503',
 'xenial-20160525',
 'xenial-20160629',
 'xenial-20160706',
 'xenial-20160713',
 'xenial-20160809',
 'xenial-20160818',
 'xenial-20160914',
 'xenial-20160923.1',
 'xenial-20161010',
 'xenial-20161114',
 'xenial-20161121',
 'xenial-20161213',
 'xenial-20170119',
 'xenial-20170214',
 'xenial-20170410',
 'xenial',
 'yakkety-20160708',
 'yakkety-20160717',
 'yakkety-20160806.1',
 'yakkety-20160826',
 'yakkety-20160919',
 'yakkety-20160923.1',
 'yakkety-20161013',
 'yakkety-20161104',
 'yakkety-20161121',
 'yakkety-20161213',
 'yakkety-20170104',
 'yakkety-20170224',
 'yakkety-20170327',
 'yakkety',
 'zesty-20161129.1',
 'zesty-20161212',
 'zesty-20170118',
 'zesty-20170224',
 'zesty-20170411',
 'zesty']
```

**Comment by gmkurtzer** [comment]
Can we extend that functionality to the cli? maybe something like:

```
singularity shell docker://ubuntu:<enter>
```
or if you just typed a few chars, show what it matches?

**Comment by vsoch** [comment]
yeah let me see if I can figure out how to do this - based on the name of the folder I think he learned [from here](https://www.gnu.org/software/bash/manual/bash.html#Programmable-Completion)

**Pull request closed without merging**